ci/tests/puppeteer/scenarios/oidc-authzcode-refresh-token-tgt-expired/script.js

```diff
@@ -0,0 +1,85 @@
+const puppeteer = require('puppeteer');
+const cas = require('../../cas.js');
+const assert = require('assert');
+
+(async () => {
+    const browser = await puppeteer.launch(cas.browserOptions());
+    const page = await cas.newPage(browser);
+
+    let url = `https://localhost:8443/cas/oidc/authorize?response_type=code&client_id=client&scope=openid%20offline_access&prompt=login&redirect_uri=https://github.com/apereo/cas&nonce=3d3a7457f9ad3&state=1735fd6c43c14`;
+
+    await cas.log(`Navigating to ${url}`);
+    await cas.goto(page, url);
+    await page.waitForTimeout(1000);
+    await cas.loginWith(page);
+    await page.waitForTimeout(1000);
+
+    if (await cas.isVisible(page, "#allow")) {
+        await cas.click(page, "#allow");
+        await page.waitForNavigation();
+    }
+
+    let code = await cas.assertParameter(page, "code");
+    await cas.log(`OAuth code ${code}`);
+
+    let accessTokenParams = `client_id=client&`;
+    accessTokenParams += "client_secret=secret&";
+    accessTokenParams += "grant_type=authorization_code&";
+    accessTokenParams += `redirect_uri=https://github.com/apereo/cas`;
+
+    let accessToken = null;
+    let refreshToken = null;
+
+    let accessTokenUrl = `https://localhost:8443/cas/oidc/token?${accessTokenParams}&code=${code}`;
+    await cas.doPost(accessTokenUrl, "", {'Content-Type': "application/json"},
+        res => {
+            
+            assert(res.data.access_token !== null);
+            assert(res.data.refresh_token !== null);
+
+            accessToken = res.data.access_token;
+            refreshToken = res.data.refresh_token;
+
+            cas.log(`Received access token ${accessToken}`);
+            cas.log(`Received refresh token ${refreshToken}`);
+        },
+        () => {
+            throw `Operation failed to obtain access token`;
+        });
+
+    assert(accessToken != null, "Access Token cannot be null");
+    assert(refreshToken != null, "Refresh Token cannot be null");
+
+    cas.log('Lets wait 10s for the TGT to expire, the RT should not expire based on its specific service expiration policy');
+    await page.waitForTimeout(10000);
+
+    let refreshTokenParams = "scope=openid%offline_access";
+    refreshTokenParams += `&grant_type=refresh_token&refresh_token=${refreshToken}`;
+
+    let refreshTokenUrl = `https://localhost:8443/cas/oidc/token?${refreshTokenParams}`;
+    cas.log(`Calling endpoint: ${refreshTokenUrl}`);
+
+    let value = `client:secret`;
+    let buff = Buffer.alloc(value.length, value);
+    let authzHeader = `Basic ${buff.toString('base64')}`;
+    await cas.log(`Authorization header: ${authzHeader}`);
+
+    await cas.doPost(refreshTokenUrl, "", {
+        'Content-Type': "application/json",
+        'Authorization': authzHeader
+    },
+        res => {
```

Comment:
Please remove all puppeteer tests and rework one of the existing tests. This test seems like it duplicates a lot of the setup and configuration.

Reply:
If you decide to keep the test around, it needs to be sufficiently more complicated. For example,

1. You have two RPs, one with a policy for RTs and one without.
2. One RT disables the tracking mechanism, and one uses the global server behavior.

For each RP, the test scenario should show that tracking mechanism works, and the RP policy works correctly when defined for the service and when it's global.

Reply:
In fact, I especially designed this new test not to mess with the one I used as a base: `oidc-authzcode-refresh-token`.

`oidc-authzcode-refresh-token` is focused on testing the link between the clients and their RTs.

My test has only one client to focus on the TGT expiration vs the RT expiration.

Reply:
This is why I'd like to keep my test as simple as it is (only one client and not two).

That said, anything is possible. I can merge both tests and test anything (link RT <-> client and TGT vs expiration at the same time), but I don't think this would be better.

Reply:
I understand, sure. It makes sense to keep the test around, but I propose that it needs to at least test with two RPs to demonstrate the difference between each policy, especially if the policy per service is configured to affect tracking. Testing with just one thing that does one thing is not too realistic.

Reply:
If you prefer that I test 2 RPs, I think it's better to merge both tests and test at the same time the link between RT and clients and the exp policies.

Reply:
Sure, that's OK with me too. Thank you!

Reply:
OK. Will do.